include/envoy/stream_info/filter_state.h

```diff
@@ -29,20 +32,42 @@ class FilterState {
   /**
    * @param data_name the name of the data being set.
    * @param data an owning pointer to the data to be stored.
-   * Note that it is an error to call setData() twice with the same data_name; this is to
-   * enforce a single authoritative source for each piece of data stored in FilterState.
+   * @param state_type indicates whether the object is mutable or not.
+   *
+   * Note that it is an error to call setData() twice with the same
```

Comment:
Just to be clear, do we (intentionally) allow overwriting mutable data with read only state? I don't have strong opinion on it but I doubt there is a use case for that.

Reply:
I lean against this--I'd recommend instead that it's always an error to cross the streams.  Flexibility does hurt if it makes it more likely people will use an interface incorrectly, and the easy things to imagine happening because of this are mistakes.  But I'll leave it up to you.

(As a general rule, I think simpler interfaces should be preferred over simpler implementations--the interface will be used many more times than the implementation will be written or rewritten, so simpler interface at the cost of more complex implementation ==> fewer bugs.  I fully acknowledge that different people can have different opinions on what interface is simpler, though :-}.)

Reply:
Fair point. I changed it so that overwriting mutable with readonly throws exception